### DISCUSSION TITLE
Issue with kong.request.get_header and ngx.var.http_xxx returning nil in Kong 3.7

### DISCUSSION THREAD

**Discussion by zhou-cl** [Kong/kong#13541] [Help]
Hi,

I’m experiencing an issue with Kong 3.7 where headers accessed via kong.request.get_header and ngx.var.http_xxx in the access phase are returning nil. As a result, when I try to set these headers using kong.service.request.set_header, they remain nil.

Could anyone explain why this might be happening? Are there specific handling steps I need to follow to resolve this issue?

Thank you!

**Comment by zhou-cl** [comment]
```lua
function axasd:access(conf)
    local bxxxx =  kong.request.get_header("axxxx")
    local c = "12312312312"
    kong.service.request.set_header('axxxx',c)
end

```

**Comment by zhou-cl** [comment]
@bungle @guanlan help

**Comment by zhou-cl** [comment]
In Kong 3.7, I am using the following Lua code:

```lua
local headers = ngx.req.get_headers()
local headerxxx= headers["xxx"]  
kong.service.request.set_header('xxxx', "111")
```
This works fine. However, when I change it to:

```lua
local headerxxx = ngx.var.http_xxx
kong.service.request.set_header('xxxx', "111")
```

it does not work. I would like to understand why this happens. Please explain the internal mechanism of Kong 3.7. In Kong 1.5.1, the previous code works correctly.

**Reply by chobits** [reply]
> As a result, when I try to set these headers using kong.service.request.set_header, they remain nil.

what's the meaning of `they remain nil`?

Does it mean your client did'nt see the set header or you could find the value from source like using `ngx.log(.... some header value...`?

**Reply by zhou-cl** [reply]
```lua
local jwtJti="111"
local headers = ngx.req.get_headers()
local a = headers["a"]
kong.service.request.set_header('b', jwtJti)  ----111
kong.service.request.set_header('a', jwtJti)---- 111
```


```lua
local jwtJti="111"
local a =  kong.request.get_header("a")
kong.service.request.set_header('b', jwtJti)  ----111
kong.service.request.set_header('a', jwtJti) ---- nil
```

**Reply by zhou-cl** [reply]
```lua
local kong = kong
local jwt_decoder = require "kong.plugins.jwt.jwt_parser"  
local utils = require "kong.tools.utils"  
local kong_meta = require "kong.meta"



local PupuJwt2HeaderHandler = {  
  PRIORITY = 100000,  
  VERSION = kong_meta.version,
}  
  
local function starts_with(str, start)
    return str:sub(1, #start) == start
end

function getppuserid()
    local jwtJti = nil
    local token = kong.request.get_header("Authorization")
    if token then
        local parts = {}
        for v in token:gmatch("%S+") do
            table.insert(parts, v)
        end
        if #parts ~= 2 then
            return nil
        end
        if parts[1]:lower() == "bearer" then
            token = parts[2]

            if not starts_with(token,"e") then
                return nil
            end

            local jwt, err = jwt_decoder:new(token)
            if err then
                kong.log.warn("failed to decode jwt: ", tostring(err), " token: ", token)
                return nil
            end
            if jwt.claims["jti"] then
                jwtJti = jwt.claims["jti"]
            end
        end
    end
    return jwtJti
end

function PupuJwt2HeaderHandler:access(conf)

    local headerPPUserid =  kong.request.get_header("pp-userid")

    local jwtJti = getppuserid()
    if headerPPUserid then
        if headerPPUserid ~= '' and jwtJti and headerPPUserid ~= jwtJti then
            kong.log.warn("jwt.jti not equal to pp-userid: jwt.jti: ", jwtJti, " <----> header: ", headerPPUserid)
        end
        
        if conf.override and jwtJti then
            kong.service.request.set_header('pp-userid',jwtJti)
        end
    else
        if jwtJti then
            kong.service.request.set_header('pp-userid',jwtJti)
        end
    end
end

return PupuJwt2HeaderHandler

**Reply by zhou-cl** [reply]
I encounter this situation when customizing plugins, but not when using pre function or post function

**Reply by chobits** [reply]
maybe I'll try some minimal code to reproduce it, If having results, i'll get back to you

**Reply by chobits** [reply]
I could not reproduce your situation, here is my test for this part

> local jwtJti="111"
> local a =  kong.request.get_header("a")
> kong.service.request.set_header('b', jwtJti)  ----111
> kong.service.request.set_header('a', jwtJti) ---- nil

From my captured package to upstream, i found `a: 111` header has been set and it's not nil

* my plugin with your code

```
function MyPluginHandler:access(conf)
  local jwtJti="111"
  local a =  kong.request.get_header("a")
  kong.service.request.set_header('b', jwtJti)  ----111
  kong.service.request.set_header('a', jwtJti) ---- nil
end
```

* added route / service / plugin to kong, we add a route to route `/mock` request to upstream `httpbin.org`

```
curl -i -s -X POST http://localhost:8001/services \
 --data name=example_service \
 --data url='http://httpbin.org'

# plugin:https://github.com/Kong/kong-ee/pull/9838/files
curl -is -X POST http://localhost:8001/services/example_service/plugins \
    --data 'name=bchen-hello'

curl -i -X POST http://localhost:8001/services/example_service/routes \
 --data 'paths[]=/mock' \
 --data name=example_route
```

* what request i send from my client curl, and note that `a: modified-me` header has been replaced by the plugin

```
curl localhost:8000/mock -o /dev/null -sv -H 'a: modified-me'
```

* captured data flow from kong to upstream (httpbin.org), you could see that `a: 111` header has been set successfully.

```
$ sudo tcpdump -i any dst host httpbin.org -s 0 -A
...
,....(5.GET / HTTP/1.1
Via: 1.1 kong/3.9.0
Host: httpbin.org
Connection: keep-alive
X-Forwarded-For: 127.0.0.1
X-Forwarded-Proto: http
X-Forwarded-Host: localhost
X-Forwarded-Port: 8000
X-Forwarded-Path: /mock
X-Forwarded-Prefix: /mock
X-Real-IP: 127.0.0.1
X-Kong-Request-Id: eb24d1213388fd1fbed3be10cef2ef27
User-Agent: curl/8.4.0
Accept: */*
a: 111
b: 111
```

**Reply by zhou-cl** [reply]
What if we make the request this way, where the original header is already nil: curl localhost:8000/mock -o /dev/null -sv?"

**Reply by chobits** [reply]
> What if we make the request this way, where the original header is already nil: curl localhost:8000/mock -o /dev/null -sv?"

same result from my test, still could proxy `a: 111` header to upstream

**Reply by zhou-cl** [reply]
I’m not sure where the issue lies:
```

local headers = ngx.req.get_headers()
local headerxxx = headers["xxx"]
kong.service.request.set_header('xxxx', "111")
```
This works fine. However, when I change it to:
```

local headerxxx = ngx.var.http_xxx
kong.service.request.set_header('xxxx', "111")
```
It does not work. I’m using the following log_format to check the logs:

```
log_format main '{'
    '"a": "$http_a",'
'}';
```
I’ve added this print statement in the post-function:


```
local a = kong.request.get_header("a")
if a ~= nil and a ~= '' then
    kong.response.set_header('response-a', a)
else
    kong.response.set_header('response-a', "a is null")
end
```
Can you help me identify the problem? I’m currently out of ideas.

**Reply by chobits** [reply]
nginx's log format $http_a may only log the original request header, I guess. You could try to capture the proxied request to see if the header is added or modified.

**Reply by zhou-cl** [reply]
But the two different methods yield different results. It’s quite frustrating.

**Reply by chobits** [reply]
Debugging this issue may be challenging and might require in-depth debugging in your own environment. I recommend using in-depth debugging methods, such as Nginx's debug logs or GDB, to help pinpoint the problem.

**Reply by zhou-cl** [reply]
thanks
